File: 206. Reverse Linked List/my_code_iter_next_node.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        
        ListNode* ReverseNodes = nullptr; // в цикле обязательно должно быть обновление Node для ReverseNodes

        while(head != nullptr){

            // применене константности будет тут ошибкой! Это указатель на НЕ реверсивную часть списка
            ListNode* NextNode = head->next; // далее эта связь будет разорвана, необходимо ее сохранить

            head->next = ReverseNodes; // перебрасываем link на иную Node с целью разворота
            ReverseNodes = head; // обновляем указатель на ответ сдвигая его

            head = NextNode; // в цикле обязательно должно быть обновление head
        }

        return ReverseNodes;

    }
};

File: 206. Reverse Linked List/my_code_iter_next_node.go
// Definition for singly-linked list.
// type ListNode struct {
//     Val  int
//     Next *ListNode
// }

/*
Time complexity: O(n)
Space complexity: O(1)
*/
func reverseList(head *ListNode) *ListNode {
    var reversed *ListNode = nil // будет хранить перевернутый список

    for head != nil {
        nextNode := head.Next   // сохраняем следующую ноду, чтобы не потерять остаток списка
        head.Next = reversed    // меняем направление ссылки
        reversed = head         // обновляем указатель на перевернутую часть
        head = nextNode         // переходим к следующей ноде
    }

    return reversed
}
